Docs Note Added to Models and Errors - Naming Collisions (#9502)
pydantic/pydantic: docs/errors/validation_errors.md

    #> 'none_required'
```

> !!! note

Reviewer: I recall that recently we added more intuitive error handling for this case - is this error still raised?

Author: I tried some examples, and it seems that the intuitive error handling might not have applied to the `none_required` error? (example below) 

**Case 1: Type Naming Clash**

```
from typing import Optional
from pydantic import BaseModel

class M1(BaseModel):
    int: Optional[int] = None

m = M1(int=123) 
'''
1 validation error for M1
int
  Input should be None [type=none_required, input_value=123, input_type=int]      
    For further information visit https://errors.pydantic.dev/2.7/v/none_required 
'''
```

**Case 2: Imported Library**

For (Field-related?) cases, a descriptive error message is raised.
```
from datetime import date
from pydantic import BaseModel, Field

class MyModel(BaseModel):
    date: date = Field()
'''
raise PydanticUserError(
pydantic.errors.PydanticUserError: Error when building FieldInfo from annotated attribute. 
Make sure you don't have any field name clashing with a type annotation
'''
```
For others, seems harder to decipher (though there's probably not much one can do to raise a specific, descriptive error message for the below example). 
```
from pydantic import BaseModel
import uuid

class Foo(BaseModel):
    uuid: uuid.UUID | None = None
'''
AttributeError: 'NoneType' object has no attribute 'UUID'
'''
```

Based on the error messages I was seeing from the above examples, I was thinking that it might be the most feasible/helpful to add a small blurb to the `none_required` portion of the validation error docs. Feel free to lmk your thoughts on that!

Reviewer: @dAIsySHEng1,

Great job on researching the different cases here. This is super helpful to have as a reference.

Also, just a small tip, if you use "```py" instead of "```" to start your code blocks, you'll get some nice syntax highlighting 🚀!

Ahh, case 2 is what I was thinking would happen. You can see some awesome docs added for that case by @Viicos [here](https://github.com/pydantic/pydantic/pull/8243/files).

So you're right, I think we want to focus on adding docs for cases 1 and 3 in this PR.

@Viicos, do you think we can in some way extend the `PydanticUserError` logic to cases 1 and 3 above, in which the field construct isn't explicitly used?

Reviewer: @dAIsySHEng1,

Your docs updates look great. I'm happy to merge approximately as is, but I'm just curious to see if we can add more robust / explicit warnings and errors :).

    You may encounter this error when there is a naming collision in your model between a field name and its type. More specifically, this error is likely to be thrown when the default value of that field is `None`.

>     For example, the following would yield the `none_required` validation error since the field  `int` is set to a default value of `None` and has the exact same name as its type, so both are interpreted to be `None`:

Reviewer: ```suggestion
    For example, the following would yield the `none_required` validation error since the field `int` is set to a default value of `None` and has the exact same name as its type, which causes problems with validation.
```
